crates/ocg_common/src/network/server.rs

```diff
@@ -0,0 +1,122 @@
+//! The network server protocol implementation, hosting a game for zero or more clients.
+
+use std::sync::Arc;
+
+use capnp_rpc::pry;
+use ocg_schemas::dependencies::capnp::capability::Promise;
+use ocg_schemas::dependencies::capnp::Error;
+use ocg_schemas::dependencies::kstring::KString;
+use ocg_schemas::schemas::network_capnp as rpc;
+use ocg_schemas::schemas::network_capnp::authenticated_server_connection::{
+    BootstrapGameDataParams, BootstrapGameDataResults, SendChatMessageParams, SendChatMessageResults,
+};
+
+use crate::network::PeerAddress;
+use crate::GameServer;
+
+/// An unauthenticated RPC client<->server connection handler on the server side.
+pub struct Server2ClientEndpoint {
+    server: Arc<GameServer>,
+    peer: PeerAddress,
+}
+
+/// An authenticated RPC client<->server connection handler on the server side.
+pub struct AuthenticatedServer2ClientEndpoint {
+    server: Arc<GameServer>,
+    peer: PeerAddress,
+    username: KString,
+    connection: rpc::authenticated_client_connection::Client,
+}
+
+impl Server2ClientEndpoint {
+    /// Constructor.
+    pub fn new(server: Arc<GameServer>, peer: PeerAddress) -> Self {
+        Self { server, peer }
+    }
+
+    /// The server this endpoint is associated with.
+    pub fn server(&self) -> &Arc<GameServer> {
+        &self.server
+    }
+
+    /// The peer address this endpoint is connected to.
+    pub fn peer(&self) -> PeerAddress {
+        self.peer
+    }
+}
+
+impl rpc::game_server::Server for Server2ClientEndpoint {
+    fn get_server_metadata(
+        &mut self,
+        _params: rpc::game_server::GetServerMetadataParams,
+        mut results: rpc::game_server::GetServerMetadataResults,
+    ) -> Promise<(), Error> {
+        let title = "OCG Server";
+        let subtitle = "Subtitles to be implemented!";
+        let mut meta = results.get().init_metadata();
+        let mut ver = meta.reborrow().init_server_version();
+        ver.set_major(0);
+        ver.set_minor(0);
+        ver.set_patch(1);
+        ver.set_build("todo");
+        ver.set_prerelease("");
```

Comment:
maybe these should be `const`s somewhere?

Reply:
Done